keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md

```diff
@@ -0,0 +1,226 @@
+This means that both the kube-controller-manager or the cloud-controller-manager can run the
+cloud specific control loops for a given cluster. For a highly available control plane to migrate
+from running only the kube-controller-manager to running both the kube-controller-manager and
+the cloud-controller-manager requires that only one process in the control plane is running
+the cloud specific controllers. This becomes non-trivial when introducing the cloud-controller-manager
+that runs overlapping controllers with the kube-controller-manager.
+
+For environments that can tolerate downtime, it will be recommended to shutdown the control plane,
+reconfigure components to use the cloud-controller-manager, and then bring the control plane back up.
+This ensures that only 1 component can be running the set of cloud controllers. For environments
+that have stricter requirements for uptime, some level of coordination is required between the two
+components to ensure that upgrading control planes does not result in running the same controller
+in more than one place while also accounting for version skew.
+
+In order to coordinate cloud-specific controllers across the kube-controller-manager and the
+cloud-controller-manager, there will be a shared resource lock between the two during leader election.
+For the cloud-controller-manager, this requires no changes since it already locks on a resource named
+`cloud-controller-manager`. However, the `kube-controller-manager` should also lock on a resource
+named `cloud-controller-manager` for the cloud specific controllers so that only 1 component in the
+cluster can run them. This requires splitting up the kube-controller-manager’s run func into
+two categories: “core” and “cloud”.
+
+<br/>
+<br/>
+<br/>
+
+![current state of leader election in kube-controller-manager](/keps/sig-cloud-provider/images/pre-kube-controller-manager-leader-election.png)
+<br/>
+**Figure 1**: current state of leader election in kube-controller-manager
+
+<br/>
+<br/>
+<br/>
+
+![proposed changes to leader election in kube-controller-manager](/keps/sig-cloud-provider/images/post-kube-controller-manager-leader-election.png)
+<br/>
+**Figure 2**: proposed changes to leader election in kube-controller-manager
+
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Splitting the Main Run Func
+
+The main run func will be split by separating the `StartControllers` function of the
+kube-controller-manager into two functions: `StartCoreControllers` and `StartCloudControllers`.
+Both functions will receive the same controller context so that the same client set and
+shared informers can be used. The call to `leaderelection.RunOrDie()` which groups the
+cloud-specific controllers will run in a go routine. The call to `leaderelection.RunOrDie()`
+which groups the core controllers will be blocking.
+
+#### Process Handling when Lock is Lost
```

Comment:
As discussed previously with @cheftako and @mcrute, I think handling process termination will be the trickiest part. Should a leader election loss in the CCM warrant the entire kube-controller-manager process to crash?